_posts/2024-11-12-a-kernel-story7.md

```diff
@@ -0,0 +1,156 @@
+date:       2024-11-12 10:00:00 +0100
+categories: kernel
+---
+
+# Introduction
+
+As mentioned in the [previous post](/kernel/2024/07/13/a-kernel-story6.html), the Chip Select (CS) signal coming from the Raspberry Pi is active-low, while the display needs an active-high signal. It took me a long time (about three months) to understand the intricacies of this configuration for the Linux Kernel and the Raspberry Pi. I won't delve into the long journey it took me to arrive at the knowledge I needed to change the signal, but I will explain what needs to be done to set the CS signal to active-high.
+
+<div style="text-align: center;" markdown="1">
+| ![Github Actions](/assets/figures/2024-11-12-Setting_CS_to_active_high_1.jpg) | 
+|:--:| 
+| *Generated by Gemini AI* |
+</div>
+
+# Changing the device-tree (DT) overlay
+
+The necessary configuration to operate this change occurs in the DT, and this is what I am going to explain in this section.
+
+## Explanation
+
+To invert the CS signal we need to set the `cs-gpio` to `ACTIVE_HIGH` in the controller **AND** set the `spi-cs-high` in the peripheral property. The kernel documentation defines a table where this is explained<sup>[1](#references)</sup>. Let me quote it:
+
+```
+There is a special rule set for combining the second flag of an cs-gpio with the optional spi-cs-high flag for SPI slaves.
+
+Each table entry defines how the CS pin is to be physically driven (not considering potential gpio inversions by pinmux):
+
+device node     | cs-gpio       | CS pin state active | Note
+================+===============+=====================+=====
+spi-cs-high     | -             | H                   |
+-               | -             | L                   |
+spi-cs-high     | ACTIVE_HIGH   | H                   |
+-               | ACTIVE_HIGH   | L                   | 1
+spi-cs-high     | ACTIVE_LOW    | H                   | 2
+-               | ACTIVE_LOW    | L                   |
+```
+
+We are interested in the third line of the table since it is the one that sets the CS pin to `ACTIVE_HIGH`. Therefore, and as mentioned above, we need to set both the `cs-gpio` to `ACTIVE_HIGH` and set the `spi-cs-high` property.
+
+## Hands-on
+
+In the Fedora distribution, overlays are located in the `/boot/efi/overlays` folder. The files in this folder are in a format understood by the Linux kernel, so you need to convert them to a human-readable format first. As an example, you can convert `spi0-1cs` with the following command:
+
+```console
+dtc -I dtb -O dts spi0-1cs.dtbo -o spi0-1cs.dtbs
+```
+
+However, I wouldn't recommend you to do this transformation, but to download the file from the Raspberry Pi repositories<sup>[2](#references)</sup>, because the transformation loses some details and the final result could be wrong.
+
+Now let's focus on the changes. First, you need to set the `cs-gpios` to `ACTIVE_HIGH` (or 0) in the `fragment@1` `__overlay__`. Example:
+
+```
+fragment@1 {
+    target = <&spi0>;

```

Comment:
probably also worth mention that the `target =` specifies the device tree node where this fragment is going to be applied once the DTBO is loaded.

Reply:
Done